starknetup.sh

```diff
@@ -0,0 +1,189 @@
+			exit 0
+			;;
+		*)
+			err "invalid option '$arg'. For more information, try '--help'."
+			;;
+		esac
+	done
+
+	assert_dependencies
+	assert_not_installed "scarb" $SCARB_UNINSTALL_DOCS
+	install_latest_asdf_plugin "scarb"
+	install_latest_version "scarb"
+	set_global_latest_version "scarb"
+
+	assert_not_installed "starknet-foundry" $STARKNET_FOUNDRY_UNINSTALL_DOCS
+	install_latest_asdf_plugin "starknet-foundry"
+
+	# Reinstall to ensure the latest version of USC is installed
+	uninstall_latest_version "starknet-foundry"
+	install_latest_version "starknet-foundry"
+	set_global_latest_version "starknet-foundry"
+
+	say "Installation complete"
+}
+
+assert_dependencies() {
+	need_cmd curl
+	need_cmd git
+	if ! check_cmd asdf; then
+		install_asdf_interactively
+	fi
+}
+
+assert_not_installed() {
+	local tool="$1"
+	local uninstall_docs_url="$2"
+
+	if ! asdf which "$tool" >/dev/null 2>&1; then
+		if check_cmd "$tool"; then
+			err "$tool is already installed outside of asdf. Please uninstall it and re-run this script. For uninstallation instructions, refer to $uninstall_docs_url."
+		fi
+	fi
+}
+
+install_latest_asdf_plugin() {
+	local plugin="$1"
+	if asdf plugin list | grep -q "$plugin"; then
+		ensure asdf plugin update "$plugin"
+	else
+		ensure asdf plugin add "$plugin"
+	fi
+}
+
+install_latest_version() {
+	local tool="$1"
+	ensure asdf install "$tool" latest
+}
+
+uninstall_latest_version() {
+	local tool="$1"
+	local latest_version
+	latest_version=$(asdf latest "$tool")
+
+	if asdf list "$tool" "^${latest_version}$" >/dev/null 2>&1; then
+		ensure asdf uninstall "$tool" "$latest_version"
+	fi
+}
+
+set_global_latest_version() {
+	local tool="$1"
+	ensure asdf global "$tool" latest
+}
+
+say() {
+	printf "starknetup: %s\n" "$1"
+}
+
+err() {
+	say "$1" >&2
+	exit 1
+}
+
+need_cmd() {
+	if ! check_cmd "$1"; then
+		err "need '$1' (command not found)"
+	fi
+}
+
+check_cmd() {
+	command -v "$1" >/dev/null 2>&1
+}
+
+ensure() {
+	if ! "$@"; then err "command failed: $*"; fi
+}
+
+install_asdf_interactively() {
+	local _profile
+	local _pref_shell
+	local _asdf_path="$HOME/.asdf"
+
+	case ${SHELL:-""} in
+	*/zsh)
+		_profile=$HOME/.zshrc
+		_pref_shell=zsh
+		;;
+	*/ash)
+		_profile=$HOME/.profile
+		_pref_shell=ash
+		;;
+	*/bash)
+		if [ "$(uname)" = "Darwin" ]; then
+			_profile=$HOME/.bash_profile
+		else
+			_profile=$HOME/.bashrc
+		fi
+		_pref_shell=bash
+		;;
+	*)
+		err "asdf-vm is required. Please install it manually and re-run this script. For installation instructions, refer to ${ASDF_INSTALL_DOCS}."
+		;;
+	esac
+
+	if [ -n "$_profile" ]; then
+		if [ ! -f "$_profile" ]; then
+			touch "$_profile"
+		fi
+
+		say "asdf-vm is required. Do you want to install it now? (y/N): "
+		read -r answer
+		case $answer in
+		[Yy]*)
+			latest_asdf_version=$(curl -s https://api.github.com/repos/asdf-vm/asdf/releases/latest | awk -F'"' '/"tag_name"/ {print $4}') || latest_asdf_version="$DEFAULT_ASDF_VERSION"
```

Comment:
I tried running this on ubuntu in docker and it failed for me (the curl request did not return anything, as it got rate limited and the default failed to assign). I've managed to make it running by:

```
                        latest_asdf_version=$(curl -s https://api.github.com/repos/asdf-vm/asdf/releases/latest | awk -F'"' '/"tag_name"
/ {print $4}')
                       latest_asdf_version=${latest_asdf_version:-"$DEFAULT_ASDF_VERSION"}
```
Can you look into this?

Reply:
if this was rate limit, maybe we could try and add retry to this query? `--fail` could probably be added as well

Reply:
> maybe we could try and add retry to this query

Generally, it's true, but this one has been rate limited on an hourly rate

> The primary rate limit for unauthenticated requests is 60 requests per hour.

Reply:
hmm, that's a nummer. your proposed solution is good enough then imo.

As a side note - maybe we should not install asdf altogether? this could be considered a dependency (and we're not installing anything curl or git). idk, just a thought.